end-user-application-machines/new-agent-vm/Install-Idle-Shutdown.sh

```diff
@@ -1,257 +1,271 @@
+        
+        saveCPU(IdleTime_now, BusyTime_now)
     
-    required_idle_time=$(( 60 * $wait_minutes))
-
-    if [ $(activeConnectionsExist) -gt 0 ]; then
-        echo exit 0
-    fi
-
-    logit "Monitoring system for idle state every $poll_interval minutes and will shutdown machine if idle (every cpu < $cpu_threshold% utilization) for $wait_minutes minutes."
-
-    startTime=0
-    # Main while loop.  Flip between two states: someone is connected, and a triggered timer
-    # counting until we shutdown or someone connects
-    while [ 1 -ne 0 ]; do
-        if [ $(activeConnectionsExist) -gt 0 ]; then
-            startTime=0
-        else
-            if [ $startTime -eq 0 ]; then
-                startTime=$(/bin/date +%s)
-            fi
-
-            cpuLoad=$(getCPULoad)
-            if [ $cpuLoad -ge $cpu_threshold ]; then
-                # a cpu load greater than threshold, assume something's going on
-                #logit "Shutdown process reset because CPU load ($cpuLoad%) > $cpu_threshold%"
-                startTime=0
-            else
-                currentTime=$(/bin/date +%s)
-                elapsedIdleTime=$(( $currentTime - $startTime ))
-
-                if [ $required_idle_time -le $elapsedIdleTime ]; then
-                logit "Shutting down idle machine."
-                sudo /sbin/shutdown now
-                fi
-            fi
-        fi
-
-        /bin/sleep $((60 * $poll_interval))
-    done 
-
-    exit 0
-}
-
-main $@
+    return round(cpu_usage, 2)
+
+def getSettings():
+    """
+    Get Settings from environment
+    """
+    # Load Values, use Defaults if needed, convert to seconds as needed
+    waitTime = float(os.environ.get('MinutesIdleBeforeShutdown', 60))*60
+    cpuThreshold = float(os.environ.get('CPUUtilizationLimit', 20))
+    
+    return (cpuThreshold, waitTime)
+
+def main():
+    """
+    Main function, monitor active sessions and CPU usage and shutdown when appropriate
+    """
+    cpuThreshold, waitTime = getSettings()
+
+    if activeConnectionsExist():
+        syslog.syslog("Session is active, not shutting down")
+        clearStartTime()
+    else:
+        cpuUsage = getCPULoad()
+        if cpuUsage < cpuThreshold:
+            startTime = getStartTime()
+            if not startTime:
+                startTime = time.time()
+                setStartTime(startTime)
+            currentTime = time.time()
+            elapsedTime = currentTime - startTime
+            syslog.syslog("CPU usage is now {cpu}%, it is below idle threshold. CPU has been idle for at least {min} minutes".format(
+                min=int(elapsedTime/60),
+                cpu=cpuUsage
+            ))
+            if elapsedTime > waitTime:
+                syslog.syslog("CPU has been idle for more than {} minutes, shutting down".format(
+                    waitTime
+                ))
+                os.system('/sbin/shutdown')
```

Comment:
`os.system()` is deprecated; this should be able to be replaced by `subprocess.Popen(['/sbin/shutdown'])`. In addition, `clearStartTime()` should be called beforehand -- otherwise, that line will never be executed because script execution will be terminated by the call to `shutdown`.